docs: update release and publishing (#529)
paritytech/substrate-api-sidecar: README.md

## Note for maintainers

> * Commits && Releases && Publishing

Reviewer: What's up with the double ampersand?

All the commits in this repo follow the [Conventional Commits spec](https://www.conventionalcommits.org/en/v1.0.0/#summary). When merging a PR, make sure 1/ to
use squash merge and 2/ that the title of the PR follows the Conventional Commits spec.

> ### Preparation

Reviewer 2: ```suggestion
### Step 1 - Preparation
```

> 1. Checkout a branch `name-v5-0-1`. When deciding what version will be released it is important to look over 1) PRs since the last release and 2) release notes for any updated polkadot-js dependencies as they may affect type definitions.

Reviewer: s/`name-v5-0-1`/`name-vMM-mm-pp`

> 1. Checkout a branch `name-v5-0-1`. When deciding what version will be released it is important to look over the past commits since the last release and the updated dependencies release notes in step 2 to see if this will be a major, minor, or patch release. If you have any questions feel free to ask the maintainers.

Reviewer 2: ```suggestion
1. Checkout a branch `name-v5-0-1`. When deciding what version will be released it is important to look over 1) PRs since the last release and 2) release notes for any updated polkadot-js dependencies as they may affect type definitions.
```

2. Ensure we have the latest polkadot-js dependencies

  * @polkadot/api [release notes](https://github.com/polkadot-js/api/releases)
  * @polkadot/apps-config [release notes](https://github.com/polkadot-js/apps/releases)
  * @polkadot/util-crypto [release notes](https://github.com/polkadot-js/common/releases)

>   Note: Every monday the polkadot-js ecosystem will usually come out with a new release. It's important that we keep up, and read the release notes for any breaking changes, or high priority updates.

Reviewer: Would be good to mention where to look for these updates.

3. After updating the dependencies, the next step is making sure the release will work against all noted runtimes for Polkadot, Kusama, and Westend. This can be handled via the [sidecar-runtime-test](https://github.com/TarikGul/sidecar-runtime-test) helper library. Instructions for how to run it are in the repos README.md. Before moving forward ensure all tests pass, and if it warns of any missing types feel free to make an issue [here](https://github.com/paritytech/substrate-api-sidecar/issues). 

4. Update the version in the package.json (this is very important for releasing on NPM).

> 5. Update `CHANGELOG.md` by looking at merged PRs since the last release. Follow the format of previous releases. 

Reviewer 3: do we got a script to get all PRs in the CHANGELOG or how's that performed?

Author: Currently we do not, we update it manually unfortunately. Earlier on with sidecar there was an issue with standard-version which started to cause inaccuracies with changelog automation. This is something we still need to fix, and probably should make priority soon. 

The solution we had was to deprecate the current changelog and start a new one. We are tracking that issue [here](https://github.com/paritytech/substrate-api-sidecar/issues/380).

Author: Fortunately though, because we release every week, there isn't a lot of PR's to actually track and log. Usually just the updated dep's, and one or two PR's.

> 5. Update the changelog by looking at the past commits since the last release. Follow the format of previous releases. 

Reviewer 2: ```suggestion
5. Update `CHANGELOG.md` by looking at merged PRs since the last release. Follow the format of previous releases. 
```

>   * Make sure to note if it has a high upgrade priority.

Reviewer 2: ```suggestion
  * Make sure to note if it is a high upgrade priority (e.g. it has type definitions for an upcoming runtime upgrade to a Parity maintained network).
```

>   * Make sure to note if it is a high upgrade priority (e.g. it has type definitions for an upcoming runtime upgrade to a Parity maintained network).

Reviewer: It would be good to keep a checklist of who to notify when we release high priority stuff (likely on to be kept on Notion, but a mention here is good also).

> 6. Commit with ex: `chore(release): 5.0.1`, then push your release up, make a PR, get review approval, then merge. 

Reviewer: What does "push your release" mean? Do you mean the `name-vx.y.z` branch?

>   * NOTE: Before pushing up as a sanity check run the 3 following commands and ensure they all run with zero errors. There is one exception with `yarn test` where you will see an `InternalServerError`, that is alright. As long as all the test suites pass. 

Reviewer: s/alright/expected/

  ```bash
  $ yarn test
  $ yarn build
  $ yarn lint
  ```

> ### Release && Pubishing to NPM and Github

Reviewer 2: ```suggestion
### Step 2 - Publish and Release to NPM and Github
```

** NPM **

NOTE: You must be a member of the @substrate NPM org and must belong to the developers team within the org. (Please make sure you have 2FA enabled.) 

1. Now that master has the commit for the release, pull down master. Run `yarn build` to ensure the target JS build is there for the NPM release. 

2. Run the following commands. (Please ensure you have 2FA enabled)

```bash
$ yarn npm login
...
$ yarn npm publish
```

** Github **

1. Make sure the tag reflects your corresponding version, and run:

```bash

> git tag v5.0.1

Reviewer 3: I'm a little annoying here but this should signed with GPG ideally ^^

```suggestion
git tag -s v5.0.1
```

Author: Nice this is good to know, TIL!

> git tag -s v5.0.1

Reviewer: s/v5.0.1/vMM.mm.pp/ where `MM` is the major, `mm` the minor and `pp` the patch release.

> git push origin --tags

Reviewer 2: ```suggestion
git push origin v5.0.1
```

Author: Ahh yes, the ole single tag push. Nice reminder, thanks!

```

> 2. Go to tags on github, inside of the repo, and click the three dots to the far right and select the option to create a release. 

Reviewer: Add a link directly?
